### ISSUE TITLE
Binary format for tracer inject/extract

### ISSUE THREAD

**Issue by ramesius** [DataDog/dd-trace-js#635]
When trying to use the `BINARY` format to inject trace data into a carrier I realised that it is not implemented in `dd-trace-js`. Would it be possible to add this as a supported format? If not some documentation on the currently supported formats :)

**Comment by rochdev** [comment]
Right now we support `TEXT_MAP` and `HTTP_HEADERS` as defined in [OpenTracing](https://doc.esdoc.org/github.com/opentracing/opentracing-javascript/variable/index.html#static-variable-FORMAT_HTTP_HEADERS), and an additional `LOG` format for log injection.

Binary format was not added as we don't have any transport which would support it as the moment.

What is your use case? I can provide an alternative way to inject in a binary protocol.

**Comment by ramesius** [comment]
My use case was the manual instrumentation of Kafka messages using headers. The headers are transported as a map `Map<string, byte[]>` (https://cwiki.apache.org/confluence/display/KAFKA/A+Case+for+Kafka+Headers) which I figured would be a good use case for the `BINARY` format.

However after some tweaking I was able to successfully instrument the payload using the `TEXT_MAP` format, just with some extra deserialisation on the receiving end :)

**Comment by rochdev** [comment]
Glad to hear that it's working for you!

Can you share how you used the `TEXT_MAP` format to propagate through Kafka for future reference?

**Comment by ramesius** [comment]
Sure! The following examples are written in typescript and use the library https://github.com/tulios/kafkajs which supports Kafka headers.

We inject tracing data into our `TEXT_MAP` carrier and assign the result to the Kafka message headers, instrumenting the message before we produce it:

```typescript
const currentSpan = tracer.scope().active();

if (currentSpan) {
 let carrier: any = {};
 tracer.inject(currentSpan, TEXT_MAP, carrier);

 if (carrier) {
  if (!message.headers) {
   message.headers = {};
  }

  Object.assign(message.headers, carrier);
 }
}

```

Then, on the consuming side we extract the DataDog tracing data from the headers, and since they are byte arrays we need to do some deserialising:

```typescript

const { headers } = message;

// Headers are consumed as buffers so need to be de-serialised
const traceId = headers['x-datadog-trace-id'].toString('utf8');
const parentId = headers['x-datadog-parent-id'].toString('utf8');
const sampled = headers['x-datadog-sampled'].toString('utf8');
const priority = headers['x-datadog-sampling-priority'].toString('utf8');

// Reconstruct the carrier
const carrier = {
 ['x-datadog-trace-id']: traceId,
 ['x-datadog-parent-id']: parentId,
 ['x-datadog-sampled']: sampled,
 ['x-datadog-sampling-priority']: priority
};

const spanContext = tracer.extract(TEXT_MAP, carrier);

```

**Comment by bensooraj** [comment]
I use Kafka too, however, I took a slightly different approach to this.

I used the inbuilt `LOG` format for extracting the `"dd": { "trace_id": 123123, "span_id": 321321 }` from the `Kafka Producer`s span context, added it to the payload and produced the message.

On the consumer side, I extracted the `dd` object from the payload received, started a new spanContext from it and used `scope.activate()` to process the message.
